### PR TITLE
disables thought updates until local push and pull are complete

### LINKED ISSUE #1071: Thoughts duplicated when typing fast
The thoughts get duplicated in the state when typing too fast. For some reason, the duplicate thought doesn't show up in the DOM but can be seen in the redux state. 
[Here's the video](https://www.awesomescreenshot.com/video/3274829?key=9dc6c6ad6cd5dfeda1da928875229116) capturing this issue.

### PR DESCRIPTION
**Pull request by anmolarora1**
fixes #1071

### CODE REVIEW

**Review by raineorshine — COMMENTED** [review]
Interesting! I'd love to understand this better. Could you explain a bit more how your solution works?

File: src/action-creators/pull.ts [hunk 1]
```diff
@@ -41,15 +42,17 @@ const pull = (contextMap: Index<Context>, { maxDepth }: PullOptions = {}): Thunk
...
-    dispatch(updateThoughts({
-      contextIndexUpdates: thoughts.contextIndex,
-      thoughtIndexUpdates: thoughts.thoughtIndex,
-      local: false,
-      remote: false,
-      // if the root is in the contextMap, force isLoading: false
-      // otherwise isLoading will not be automatically unset by updateThoughts if the root context is empty
-      ...ROOT_ENCODED in contextMap ? { isLoading: false } : null
-    }))
+    if (!hasPushes(getState()) && !getState().isPulling) {
```
[the remark this answers is not in the document]
Comment (anmolarora1) [reply src/action-creators/pull.ts:45]: You're right, the logic for having `isPulling` is flawed. Sorry for the trouble.

**Comment by anmolarora1** [comment]
@raineorshine The cause of the duplication is that because we're [updating thoughts](https://github.com/cybersemics/em/blob/a9789a7e38b42d372bf2457f3ee01b3a1f22edf2/src/action-creators/pull.ts#L44) after every `pull` operation, albeit debounced. We need to ensure that there are no multiple such calls while updating the same thought because the thought ID gets updated with every update but as [described on the issue](https://github.com/cybersemics/em/issues/1071#issuecomment-814998712), it's the async nature of the push and `pull` ops that can't sync in with the local updates.

Through this PR -

1. I was aiming to detect ongoing push and pull operations, and disable `updateThoughts` if there are any. The former has been done already, but the latter doesn't seem possible without introducing some hack because `updateThoughts` is [integrated within the pull](https://github.com/cybersemics/em/blob/a9789a7e38b42d372bf2457f3ee01b3a1f22edf2/src/action-creators/pull.ts#L44).

2. The debounce time for updatePullQueue is too short(10ms) to prevent multiple calls, so I've updated it to 100ms, and that does reduce the chances of error - *It works as long as the time to pull from local provider stays below the threshold
I understand this isn't a very reliable solution but with my tests so far, it has offered some promising results.  

With these changes alone, I haven't been able to reproduce the issue so far. But to be 100% sure, we'd need to complete the first part of it (detecting ongoing pulls and preventing thought updates if there are any). However, I believe it won't be significant anymore when we shift to a `subscription` based approach for local update (Dexie) just like the firebase (#932), which is necessary to support multiple tabs' sync. We'll add a way to distinguish the self-triggered updates from the rest ([as discussed here](https://github.com/cybersemics/em/issues/1071#issuecomment-816827411)). We can either do that as a part of this issue, or the one that involves shifting to a subscription model.
Let me know what you think.

File: src/action-creators/pull.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (anmolarora1) [reply src/action-creators/pull.ts:45]: You're right. The approach for `isPulling` wasn't correctly implemented as explained [here](https://github.com/cybersemics/em/pull/1087#issuecomment-824615621). Sorry for the trouble.

**Comment by raineorshine** [comment]
> @raineorshine The cause of the duplication is that because we're [updating thoughts](https://github.com/cybersemics/em/blob/a9789a7e38b42d372bf2457f3ee01b3a1f22edf2/src/action-creators/pull.ts#L44) after every `pull` operation, albeit debounced. We need to ensure that there are no multiple such calls while updating the same thought because the thought ID gets updated with every update but as [described on the issue](https://github.com/cybersemics/em/issues/1071#issuecomment-814998712), it's the async nature of the push and `pull` ops that can't sync in with the local updates.

Yes, I made a fatal mistake designing the reconciliation logic. I assumed it was always possible to compare timestamps to prevent old changes from overriding new changes. However I didn't fully take into account the fact that if the thought was deleted locally, there would be nothing to compare the incoming thought to and it would look like an addition.

> 1. I was aiming to detect ongoing push and pull operations, and disable `updateThoughts` if there are any. The former has been done already, but the latter doesn't seem possible without introducing some hack because `updateThoughts` is [integrated within the pull](https://github.com/cybersemics/em/blob/a9789a7e38b42d372bf2457f3ee01b3a1f22edf2/src/action-creators/pull.ts#L44).

The main question I have is, if you are disabling `updateThoughts` completely rather than selectively filtering them, how can you be sure that you are not missing any updates before a push? It seems like it may be introducing a short time period where the local state is not updated, which could potentially mean that out-of-date data gets pushed to the remote.

(I should have gone with CRDTs from the beginning. They weren't super easy to use yet when I started **em**.)

All of this may be moot once we implement subscriptions, I'm just thinking of the best way to transition. We already have some serious data integrity issues, so we have to be 100% sure that new code does not introduce additional data integrity issues.

> 2\. The debounce time for updatePullQueue is too short(10ms) to prevent multiple calls, so I've updated it to 100ms, and that does reduce the chances of error - *It works as long as the time to pull from local provider stays below the threshold
>     I understand this isn't a very reliable solution but with my tests so far, it has offered some promising results.

Your intuition is correct. Since it's async, we don't know how long it will take to return so we should not code against that. The debounce should be purely for performance reasons rather than functional reasons. Also, 100ms is significant for the user experience when we want thoughts to appear as instantaneously as possible as they navigate. I would argue that increasing the delay just avoids the problem rather than solving it.

> With these changes alone, I haven't been able to reproduce the issue so far.

Sure, I believe that. I think it solves the issue, the main problem is the gap it opens up in keeping local thoughts up-to-date.

> However, I believe it won't be significant anymore when we shift to a `subscription` based approach for local update (Dexie) just like the firebase (#932), which is necessary to support multiple tabs' sync. We'll add a way to distinguish the self-triggered updates from the rest ([as discussed here](https://github.com/cybersemics/em/issues/1071#issuecomment-816827411)). We can either do that as a part of this issue, or the one that involves shifting to a subscription model.

Yes, I think this is key. Given we don't have proper conflict resolution via CRDTs, we have to keep the local state in sync with the remote. Once we add remote/IndexedDB subscriptions, we have to assume that the local state is up-to-date.

So what's the best way to get there? We originally sidetracked from #932 in order to fix this bug, but it seems like the best approach is a full implementation of #932 for both IndexedDB and Firebase so that we can eliminate the additional pull cycle.

Thanks for discussing!

File: src/action-creators/pull.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (raineorshine) [reply src/action-creators/pull.ts:45]: Not sure I understand yet. I'll come back to it after our main discussion has a clear direction if it's still relevant.

**Comment by raineorshine** [comment]
@shresthabijay Hey Bijay! I'd love to get your input on this. I'm pulling you into a long discussion :). It's worth having another opinion since it is such a critical piece of the infrastructure.

I would start with this clear description of the problem:
https://github.com/cybersemics/em/issues/1071#issuecomment-814998712

Now our main points of discussion are:

- How do we prevent the outdated `pull` result (which causes the duplication) without preventing valid `pull` results?
- Does subscribing to data provider changes (IndexedDB and Firebase) solve this issue by ensuring the data is always up-to-date, thus allowing us to remove the additional `pull` completely? If so, should we work directly on that or should we solve this with an intermediate solution?

Thanks so much!

**Comment by anmolarora1** [comment]
> > @raineorshine The cause of the duplication is that because we're [updating thoughts](https://github.com/cybersemics/em/blob/a9789a7e38b42d372bf2457f3ee01b3a1f22edf2/src/action-creators/pull.ts#L44) after every `pull` operation, albeit debounced. We need to ensure that there are no multiple such calls while updating the same thought because the thought ID gets updated with every update but as [described on the issue](https://github.com/cybersemics/em/issues/1071#issuecomment-814998712), it's the async nature of the push and `pull` ops that can't sync in with the local updates.
> 
> Yes, I made a fatal mistake designing the reconciliation logic. I assumed it was always possible to compare timestamps to prevent old changes from overriding new changes. However I didn't fully take into account the fact that if the thought was deleted locally, there would be nothing to compare the incoming thought to and it would look like an addition.

That's correct. Maybe we can rectify that if we change our underlying architecture of storing thoughts, so every update isn't a combination of delete & add op. That'd probably improve the performance and save us from some data-integrity issues, but wouldn't completely fix them because the delete operations still need to be accounted for. So it's something for later.



> > 1. I was aiming to detect ongoing push and pull operations, and disable `updateThoughts` if there are any. The former has been done already, but the latter doesn't seem possible without introducing some hack because `updateThoughts` is [integrated within the pull](https://github.com/cybersemics/em/blob/a9789a7e38b42d372bf2457f3ee01b3a1f22edf2/src/action-creators/pull.ts#L44).
> 
> The main question I have is, if you are disabling `updateThoughts` completely rather than selectively filtering them, how can you be sure that you are not missing any updates before a push? It seems like it may be introducing a short time period where the local state is not updated, which could potentially mean that out-of-date data gets pushed to the remote.
> 
> (I should have gone with CRDTs from the beginning. They weren't super easy to use yet when I started **em**.)
> 
> All of this may be moot once we implement subscriptions, I'm just thinking of the best way to transition. We already have some serious data integrity issues, so we have to be 100% sure that new code does not introduce additional data integrity issues.


> 
> > However, I believe it won't be significant anymore when we shift to a `subscription` based approach for local update (Dexie) just like the firebase (#932), which is necessary to support multiple tabs' sync. We'll add a way to distinguish the self-triggered updates from the rest ([as discussed here](https://github.com/cybersemics/em/issues/1071#issuecomment-816827411)). We can either do that as a part of this issue, or the one that involves shifting to a subscription model.
> 
> Yes, I think this is key. Given we don't have proper conflict resolution via CRDTs, we have to keep the local state in sync with the remote. Once we add remote/IndexedDB subscriptions, we have to assume that the local state is up-to-date.
> 
> So what's the best way to get there? We originally sidetracked from #932 in order to fix this bug, but it seems like the best approach is a full implementation of #932 for both IndexedDB and Firebase so that we can eliminate the additional pull cycle.
> 
> Thanks for discussing!



CRDTs are great and would solve the problem, but the solution would simply be a by-product of the transition that we might not need right now. We'd certainly need CRDTs when we support `collaborative editing` because that brings possibilities of conflicts but right now, our objective is to maintain data integrity between multiple sessions (local & remote) for a `single user`. As far as I understand, there can't be any conflicting scenarios except one - when there're both **online** and **offline** edits to the same thought through multiple devices.

1. User types a thought `hello world` on her phone (while online)
2. User updates it to `hello Raine` on her phone (offline)
3. User opens `em` on her laptop, updates `hello world`(since that's the last synced piece of info)  to `hello Anmol` (online)
4. User switches back to her phone, gets online, and there's a conflict between `hello Raine` and `hello Anmol` 

I'd like to hear from you on the likelihood of that scenario, and I believe that'd help us choose the solution for now unless there are any other scenarios that I missed :)


Regarding the subscription logic, we can add to each session (remote/local) a unique session ID that gets passed to the sync events (through a subscription), and we'd merge only those updates that arrive from the other sessions, ignoring the local ones. 
1. When the user isn't logged in, we'd only need to sync between tabs and since the user can be active only on one tab, the other tabs would always contain the updated information.

2. When logged in(remote session), we'd be able to sync between devices in the same way - update the firebase and all other sessions would be updated through subscription. Since there would be updates from remote and local in this case, we'd need to ensure that there are no conflicts between local and remote updates, so we need to choose one as a source of truth.
   Based on the assumptions 
   -  We use the same session ID for local and remote sessions - each update will include the session ID.
   - All the local session IDs get stored in the localstorage.

   If there are 4 active sessions - L1, L2, L3 on the desktop, and L4 on mobile, sessions in the desktop's localstorage would be L1, L2, & L3  and that of mobile would be just L4
   -  If L2 gets a local update from L1 or L3  - accept
   -  If L2 gets a remote update from L1 or L3 - discard
   -  If L2 gets an update from L4 - accept
   -  If L4 gets an update from L1 or L2 or L3 - accept

**Comment by raineorshine** [comment]
> That's correct. Maybe we can rectify that if we change our underlying architecture of storing thoughts, so every update isn't a combination of delete & add op. That'd probably improve the performance and save us from some data-integrity issues, but wouldn't completely fix them because the delete operations still need to be accounted for. So it's something for later.

I agree.

> CRDTs are great and would solve the problem, but the solution would simply be a by-product of the transition that we might not need right now. We'd certainly need CRDTs when we support `collaborative editing` because that brings possibilities of conflicts but right now, our objective is to maintain data integrity between multiple sessions (local & remote) for a `single user`.

Yes, although I'm not sure there is a real distinction between single user and multiple users. CRDTs seem just as useful for conflict resolution among different devices as different users. It's the same problem from what I can tell.

> As far as I understand, there can't be any conflicting scenarios except one - when there're both **online** and **offline** edits to the same thought through multiple devices.
> 
> 1. User types a thought `hello world` on her phone (while online)
> 2. User updates it to `hello Raine` on her phone (offline)
> 3. User opens `em` on her laptop, updates `hello world`(since that's the last synced piece of info)  to `hello Anmol` (online)
> 4. User switches back to her phone, gets online, and there's a conflict between `hello Raine` and `hello Anmol`
> 
> I'd like to hear from you on the likelihood of that scenario, and I believe that'd help us choose the solution for now unless there are any other scenarios that I missed :)

Yes, that's a good example! That is a common scenario. 

A slight variation also is editing a thought one device and editing a descendant on another. Due to edits changing the entire descendant tree, this could introduce a conflict/duplicate. 

However, I don't believe it's related to the duplication issue #1071, which involves timing issues between local state and remote in a single online device.

> Regarding the subscription logic, we can add to each session (remote/local) a unique session ID that gets passed to the sync events (through a subscription), and we'd merge only those updates that arrive from the other sessions, ignoring the local ones.

I'm going to move discussion of a subscription implementation to #932. For our purposes here, we mainly need to determine if subscription support 1) solves #1071, 2) do we need an intermediate solution like this PR, and 3) does the intermediate solution introduce additional data integrity risks. In other words,

> Does subscribing to data provider changes (IndexedDB and Firebase) solve this issue by ensuring the data is always up-to-date, thus allowing us to remove the additional pull completely? If so, should we work directly on that or should we solve this with an intermediate solution?

**Comment by shresthabijay** [comment]
@raineorshine @anmolarora1 I am into this dicussion very late. I understand the issue #1071 is due to pull lagging, causing the stale state to be merged to an already updated state just like @anmolarora1 mentioned.  We are constantly pulling the local changes because we need to support editing in multiple tabs on the same device. 

I like the @anmolarora1 approach to subscription logic https://github.com/cybersemics/em/pull/1087#issuecomment-825488407. However, it doesn't guarantee to solve the timing issue. 

In the case of local editing with multiple tabs, we can introduce a `lastUpdated` to the local DB that represents the timestamp when the thoughts were last updated. We also introduce `lastSyncLocal` to the state.  It's just the `lastUpdated` copied when the state is last synced to local db.  Subscribed sessions (using dexie-observable) in same device will get notified and check `lastSyncLocal` against local DB `lastUpdated`. Only pull and apply the changes to the state if `lastSyncLocal` is outdated. In case of ongoing push, we cancel the DB transaction and merge the incoming changes and then push the changes. I think this will solve any timing issue.

**Comment by shresthabijay** [comment]
> Yes, although I'm not sure there is a real distinction between single user and multiple users. CRDTs seem just as useful for conflict resolution among different devices as different users. It's the same problem from what I can tell.
> 
> >

Agreed.

**Comment by raineorshine** [comment]
> In the case of local editing with multiple tabs, we can introduce a `lastUpdated` to the local DB that represents the timestamp when the thoughts were last updated. We also introduce `lastSyncLocal` to the state. It's just the `lastUpdated` copied when the state is last synced to local db. Subscribed sessions (using dexie-observable) in same device will get notified and check `lastSyncLocal` against local DB `lastUpdated`. Only pull and apply the changes to the state if `lastSyncLocal` is outdated. In case of ongoing push, we cancel the DB transaction and merge the incoming changes and then push the changes. I think this will solve any timing issue.

Which local would `lastSyncLocal` be referring to? There can be any number of devices logged in and connected to the remote.

Also, I think we need per-thought `lastUpdated`. A single `lastUpdated` value for all of the state does not take into account the fact that different thoughts may get loaded at different times, depending on where the user is navigating.

> I like the @anmolarora1 approach to subscription logic [#1087 (comment)](https://github.com/cybersemics/em/pull/1087#issuecomment-825488407). However, it doesn't guarantee to solve the timing issue. I would like to add something to his approach.

Thanks for pointing this out. Could you explain how a subscription implementation won't fully solve #1071? We were under the assumption that adding subscription to both IndexedDB and Firebase would allow us to remove the `pull` that happens before each `push` and thus eliminate the duplication problem.

**Comment by shresthabijay** [comment]
> > In the case of local editing with multiple tabs, we can introduce a `lastUpdated` to the local DB that represents the timestamp when the thoughts were last updated. We also introduce `lastSyncLocal` to the state. It's just the `lastUpdated` copied when the state is last synced to local db. Subscribed sessions (using dexie-observable) in same device will get notified and check `lastSyncLocal` against local DB `lastUpdated`. Only pull and apply the changes to the state if `lastSyncLocal` is outdated. In case of ongoing push, we cancel the DB transaction and merge the incoming changes and then push the changes. I think this will solve any timing issue.
> 
> Which local would `lastSyncLocal` be referring to? There can be any number of devices logged in and connected to the remote.
> 
> Also, I think we need per-thought `lastUpdated`. A single `lastUpdated` value for all of the state does not take into account the fact that different thoughts may get loaded at different times, depending on where the user is navigating. I used `state.lastUpdated` in an early implementation and then moved to per-Lexeme and per-Parent `lastUpdated` for itemized reconciliation.
> 
> > I like the @anmolarora1 approach to subscription logic [#1087 (comment)](https://github.com/cybersemics/em/pull/1087#issuecomment-825488407). However, it doesn't guarantee to solve the timing issue. I would like to add something to his approach.
> 
> Thanks for pointing this out. Could you explain how a subscription implementation won't fully solve #1071? We were under the assumption that adding subscription to both IndexedDB and Firebase would allow us to remove the `pull` that happens before each `push` and thus eliminate the duplication problem.

Now that I think of we may not run into a timing issue with indexed db subscription. I was thinking about multiple updates from different tabs when offline. But in the real scenario, even if there are multiple tabs we would only be actively editing on one of them, so the subscription should instantly notify all the tabs at once. What do you think ?

**Comment by raineorshine** [comment]
> Now that I think of we may not run into a timing issue with indexed db subscription. I was thinking about multiple updates from different tabs when offline. But in the real scenario, even if there are multiple tabs we would only be actively editing on one of them, so the subscription should instantly notify all the tabs at once. What do you think ?

Yup, I think that’s a fair assumption.

**Comment by raineorshine** [comment]
> > > The solution involves pulling thoughts for selected actions instead of every action. I'm hoping that this would be completed soon.

> > The solution needs to be generalized so that it works for any actions. This was an intentional design choice so that the system is extensible as new actions are added to the app. Feel free to blacklist certain actions, but don’t use a whitelist approach that requires new actions to be manually added.

> Here's what I think -
> Now that we have subscription-based updates to sync visible thoughts in real-time, we should consider only those actions that impact the non-visible thoughts. 
> While blacklisting is an easier option, I think an effective solution is to whitelist actions that would need a pull because that saves us from the overhead of calculating visible and pending contexts for every action - something that we're currently doing.
> Do you think that grouping actions into different categories would be helpful? Something of this sort -
> 1. UX critical actions - The actions that impact visible thoughts (editThought, moveThought/Up/Down, indent, outdent, extractThought, splitThought, etc). With real-time updates in place, we shouldn't need a manual pull in response to any of these actions.
> 2. Nav actions - actions like setCursor, cursorUp, cursorDown, cursorBack, to name a few. We need to trigger a pull for all these actions since they change visible thoughts/contexts that might've been updated in the background.
> 3. Meta actions - such as tutorial, search, alert, showModal, toggleTopControlsAndBreadcrumbs, error, dragInProgress, etc. These do not directly impact the thoughts so we probably don't need to worry about them.
> We can add more categories, or add further details to each one. This would likely help us figure out an appropriate solution for this issue and also be useful for subsequent issues like undo/redo. 
> I might be missing something here so please feel free to add.

In order to determine if the visible thoughts have changed, we have `getVisibleContexts` in the `pullQueue` which uses `expandThoughts`. `expandThoughts` is decoupled from any specific actions. There is no way to strictly know which actions will change visible thoughts or not, since the expansion logic has a variety of conditions, including meta attributes (`=pin`, `=view/Table`) and content (thoughts ending in `:`). It is better to check which thoughts are visible regardless of which actions have been called.

For actions that do not affect the cursor or visible thoughts at all, then `expandedContexts` will not change, and `updatePullQueue` will short circuit, so this path should have negligible overhead. If certain actions are a performance bottleneck, we can blacklist them from the `pullQueue` without losing the generality of the solution.

**Comment by raineorshine** [comment]
The current plan is to remove the additional pulls that are causing the duplication once we have added subscriptions that keep the data up-to-date.

**Pull request closed without merging**